### PR TITLE
Auto generated go config package.

### PR DESCRIPTION
**Pull request by ppknap**
This PR removes hard-coded static configuration from config package. Golang `config` package is now auto-generated from values set in ~~`config/goGenerate.coffee`~~ `config/generateKonfig.coffee` module.

`main.production.coffee` file should be updated with following code:

``` coffeescript
options.publicLogsS3BucketName or= 'koding-publiclogs'
options.proxySubdomain or= 'p'
```
## Motivation and Context

Store configuration in one place and allow to generate it based on ./configure options.
## How Has This Been Tested?

Unit tests for config package. (added to wercker)
~~Golden-file based tests made in order to validate go generate results. (added to wercker)~~

### CODE REVIEW

File: go/src/koding/config/base_test.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+func TestGetBucket(t *testing.T) {
+	var testTmpls = map[string]*template.Template{
+		`buckets.A`:    template.Must(template.New(`buckets.A`).Parse(`{"name":"bucket","region":"reg"}`)),
+		`buckets.Tmpl`: template.Must(template.New(`buckets.Tmpl`).Parse(`{"name":"test-{{.Environment}}","region":"reg-{{.Group}}"}`)),
+		`invalid.Tmpl`: template.Must(template.New(`invalid.Tmpl`).Parse(`{"name":"{{.Missing}}","region":"reg"}`)),
+		`invalid.Json`: template.Must(template.New(`invalid.Json`).Parse(`{"name":bucket,"region":"reg"}`)),
+	}
+
+	tests := []struct {
```
Comment (rjeczalik) [inline go/src/koding/config/base_test.go]: just a nitpick - if you replace `[]struct{` with `map[string]struct{` and label each test with a description, you could remove `// 0 //` comments + have a nice description of a test-case (`"this one should fail because...": {` etc.)

File: go/src/koding/config/genconfig.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	"strings"
+	"text/template"
+)
+
+var (
+	input  = flag.String("i", "-", "")
+	output = flag.String("o", "-", "")
+)
+
+var t = template.Must(template.New("").Parse(`// This file is auto-generated. DO NOT EDIT!
```
Comment (cihangir) [inline go/src/koding/config/genconfig.go]: i am not sure why we need this template and all the file and template generation. could you please explain? You are already generating a json file  in `goGenerate`, why not just use json encode/decode? is not it simpler? we are already generating the config for each env while deploying

File: config/goGenerate.coffee [hunk 3]
```diff
@@ -0,0 +1,18 @@
+module.exports = (options, credentials) ->
```
Comment (usirin) [inline config/goGenerate.coffee]: can we rename this to `generage-go.coffee` or `generateGo.coffee`?

File: config/main.dev.coffee [hunk 4]
```diff
@@ -84,6 +86,9 @@ Configuration = (options = {}) ->
   KONFIG.workers = require('./workers')(KONFIG, options, credentials)
   KONFIG.client.runtimeOptions = require('./generateRuntimeConfig')(KONFIG, credentials, options)
 
+    # Generate static variables for Go.
+  KONFIG.goGenerate = require('./goGenerate')(options, credentials)
```
Comment (usirin) [inline config/main.dev.coffee]: Same here.

File: config/main.sandbox.coffee [hunk 5]
```diff
@@ -123,6 +125,9 @@ Configuration = (options = {}) ->
 
   KONFIG.client.runtimeOptions = require('./generateRuntimeConfig')(KONFIG, credentials, options)
 
+    # Generate static variables for Go.
+  KONFIG.goGenerate = require('./goGenerate')(options, credentials)
```
Comment (usirin) [inline config/main.sandbox.coffee]: Same here.

File: config/generateShellEnv.coffee [hunk 6]
```diff
@@ -9,7 +9,7 @@ module.exports.create = (KONFIG, options = {}) ->
   traverse.paths(KONFIG).forEach (path) ->
     node = traverse.get KONFIG, path
     return  if typeof node is 'object'
-    add "KONFIG_#{path.join('_')}".toUpperCase(), node
+    add "KONFIG_#{path.join('_').replace(/\./g, "_")}".toUpperCase(), node
```
Comment (szkl) [inline config/generateShellEnv.coffee]: What property had a dot in its name?

File: config/goGenerate.coffee (hunk 3 above)
Comment (szkl) [inline config/goGenerate.coffee]: Module name should explain what it generates.

Comment (ppknap) [inline config/goGenerate.coffee]: @szkl you are right. I followed naming conversions I see files like `workers.cofffee` which also returns an object. According to this, what about `goConfig.coffee` or `generateGoConfig.coffee` (however the former includes a `generate` verb which suggests that we do some actions inside). I'd rather not use `generage-go.coffee` since it's inconsistent with other file names in `config` directory.

Comment (rjeczalik) [inline config/goGenerate.coffee]: generateGoConfig.coffee?

File: go/src/koding/config/genconfig.go (hunk 2 above)
Comment (ppknap) [inline go/src/koding/config/genconfig.go]: Depending on my knowledge I assume that current implementation of `koding/config` uses `go generate` and `go-bindata` to provide a way of storing build time configuration directly in binary file. This allows us to have all necessary data in a single executable file.

Since the configuration from this package is used by klient we don't have to provide additional config.json files which may become inconsistent with those used by other parts of the infrastructure (like tunnelserver etc.).

So, as you said, we can still use `go generate` + `json.Unmarshal` + `json.Marshal` + `go-bindata`. And yes, this could be even slightly easier than proposed solution. So why did I choose it over JSON-like one:

##### Simplicity:

There is no difference between this PR and JSON solution. If you want to add a new variable, you only need to edit `dontKnowHowToNameIt.coffee` file.

##### Correctness:

Let say we used JSON solution and have statement like this:

```go
endpoint := config.Builtin.Endpoints.URL("tunel", env)
```

we incorrectly wrote `tunel` instead of `tunnel`. This will compile and may even pass the tests.

Auto-generated functions enforce compiler to check this at compile time, so if we made error like above:

```go
endpoint := config.MustTunelURL(env)
```

the code won't even compile unless we have endpoint.tunel variable defined in `dontKnowHowToNameIt.coffee` file.

##### Maintainability:

Another example, when we want to remove an unused variable from existing configuration. In JSON-like solution we just need to grep the code and remove all references to string we want to remove. But, what if we have code like that:

```go
name := blackBox.GetEndpointName() // this can be everything.
endpoint := config.Builtin.Endpoints.URL(name, env)
```

without going deeper in the logic we can't be sure that removing a variable is not going break something. Static-function-like code enforces us to make such logic explicit:

```go
switch blackBox.GetEndpointName() {
	case "tunnel":
		endpoint = config.TunnelURL(env)
	// ...
}
```

so it will be easy to check if our variable is still needed or not.

TL;DR

I used templates and auto-generated files to improve code maintainability and enforce correctness of configuration variables.

File: config/generateShellEnv.coffee (hunk 6 above)
Comment (ppknap) [inline config/generateShellEnv.coffee]: I really like @cihangir 's idea proposed here #9240 to filter out unused variables. Currently, this script generates > 600 env variables which are mostly unused. I don't use variables generated by this file at all and had to add this logic only because it consumes `goGenerate` object.

`.replace(/\./g, "_")` dosn't change any of existing variables so I think it's safe to left it here.

File: config/goGenerate.coffee (hunk 3 above)
Comment (szkl) [inline config/goGenerate.coffee]: @ppknap 👍  I'm okay with these names.

> Since the configuration from this package is used by klient ...

What about generateKlientConfig.coffee?

Comment (ppknap) [inline config/goGenerate.coffee]: these variables are not only used in Klient so `generateKlientConfig` may be misleading.

File: config/generateShellEnv.coffee (hunk 6 above)
Comment (szkl) [inline config/generateShellEnv.coffee]: I've thought I've left a bug that didn't cause any issues yet :)

Well, most of the environment variables are not directly referenced because they are read into an object called KONFIG in node modules and respective struct types in golang packages during start phase of programs. It provides a way to set configuration in runtime without any configure or build process.

File: go/src/koding/config/genconfig.go (hunk 2 above)
Comment (szkl) [inline go/src/koding/config/genconfig.go]: I'm not a big fan of auto-generation but if there isn't any way to achieve correctness and maintainability in golang like you've explained then it sounds good to me.

Comment (cihangir) [inline go/src/koding/config/genconfig.go]: i still cant follow the idea, sorry :/ 

What is the problem you see creating a json/toml from the coffee and using multiconfig to read that and using `required` tag?

It is simple, it is correct, it is maintainable - we are using it in all of our workers (> 25 workers)

**Review by cihangir — CHANGES_REQUESTED** [review]
waiting for answer/clarification

File: go/src/koding/config/genconfig.go (hunk 2 above)
Comment (cihangir) [inline go/src/koding/config/genconfig.go]: >Yes, I'd probably use the solution you provided if I was creating new workers. This is completely fine.

If you could use that for new workers, you can use for the old ones as well.

Yep you are saying we have a different use case but i dont see it as different. I see it as just complex.

The complexity you create comes from the first design decision that was made by trying to use "strings" to get the configuration you are looking for. And to remediate that complexity you are adding another level of complexity by creating templates on the fly and adding them as "generated" files. Another level of complexity you add comes with "code generation" part, not to use json serialization you have implemented 477 more lines of code ( i dont count the lines usually, just stating here because previously you mentioned about LOC )

What i wanted to see with my request was simple config generation on configure step, i am not against using bindata.

>  I see no advantages in such solution.

You will have a completely build-time-safe structs and you can refactor them easily whenever you want. We had great success with using multiconfig on many places.

For the future generations lets have simple and understandable codebase.

File: config/goGenerate.coffee [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  endpoints      :
+    ip           : "https://#{options.proxySubdomain}.koding.com/-/ip"
+    ipCheck      : "https://#{options.proxySubdomain}.koding.com/-/ipcheck"
+    kdLatest     : "https://koding-kd.s3.amazonaws.com/{{.Group}}/latest-version.txt"
+    klientLatest : "https://koding-klient.s3.amazonaws.com/{{.Environment}}/latest-version.txt"
+    kloud        : "#{options.publicHostname}/kloud/kite",
+    kontrol      : "#{options.publicHostname}/kontrol/kite",
+    tunnelServer : "#{options.tunnelUrl}/kite"
+  routes         :
+    {
```
Comment (cihangir) [inline config/goGenerate.coffee]: you can omit brackets

File: config/goGenerate.coffee [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+module.exports = (options) ->
+  environment    : options.environment
+  buckets        :
+    publicLogs   :
+      name       : options.publicLogsS3BucketName
+      region     : 'us-east-1'
+  endpoints      :
+    ip           : "https://#{options.proxySubdomain}.koding.com/-/ip"
+    ipCheck      : "https://#{options.proxySubdomain}.koding.com/-/ipcheck"
+    kdLatest     : "https://koding-kd.s3.amazonaws.com/{{.Group}}/latest-version.txt"
```
Comment (cihangir) [inline config/goGenerate.coffee]: this config should be decided on configure step, not on go side.

File: config/goGenerate.coffee [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    klientLatest : "https://koding-klient.s3.amazonaws.com/{{.Environment}}/latest-version.txt"
```
Comment (cihangir) [inline config/goGenerate.coffee]: same here with previous comment

File: go/src/koding/config/genconfig.go (hunk 2 above)
Comment (ppknap) [inline go/src/koding/config/genconfig.go]: Okay, I will try to simplify this a bit.

File: config/main.default.coffee [hunk 10]
```diff
@@ -45,6 +45,8 @@ Configuration = (options = {}) ->
   options.paymentBlockDuration = 2 * 60 * 1000 # 2 minutes
   options.credentialPath or= "$KONFIG_PROJECTROOT/config/credentials.#{options.environment}.coffee"
   options.clientUploadS3BucketName or= 'kodingdev-client'
+  options.publicLogsS3BucketName or= 'kodingdev-publiclogs'
+  options.proxySubdomain or= 'dev-p2'
```
Comment (ppknap) [inline config/main.default.coffee]: yes, and it stores `dev-t.koding.com`. :)

File: config/goGenerate.coffee (hunk 8 above)
Comment (ppknap) [inline config/goGenerate.coffee]: this can be hard to achieve for environments like `sandbox` which gets their version from `development` bucket. Maybe it's better to just create `sandbox` bucket on S3 so we can indeed get rid of {{.Group}}.

File: config/goGenerate.coffee (hunk 9 above)
Comment (ppknap) [inline config/goGenerate.coffee]: the same problem as with previous one. This also includes managed/devmanaged environments which can be requested by klient.

File: config/goGenerate.coffee (hunk 8 above)
Comment (rjeczalik) [inline config/goGenerate.coffee]: > Maybe it's better to just create sandbox bucket on S3 so we can indeed get rid of {{.Group}}

I don't think it's a good idea - currently we have `production`, `development` and `managed` and `devmanaged` channels (buckets) - we should get rid at least of two managed and do not create more buckets.

**Comment by ppknap** [comment]
Much more simplified version. @cihangir @rjeczalik

File: go/src/koding/klient/info/publicip/publicip.go [hunk 11]
```diff
@@ -64,15 +63,15 @@ func (d *director) EndpointPublicIP() string {
...
+	return config.Builtin.Endpoints.IP
 }
 
 func (d *director) EndpointIsReachable(port string) string {
 	if d.fallback {
 		return "http://rjk.io/test/" + port
 	}
 
-	return config.Builtin.Endpoints.URL("ipcheck", protocol.Environment) + "/" + port
+	return config.Builtin.Endpoints.IPCheck + "/" + port
```
Comment (cihangir) [inline go/src/koding/klient/info/publicip/publicip.go]: this string interpolation can go into config as well and we can get rid of this helper function

File: go/src/koding/klient/main.go [hunk 12]
```diff
@@ -67,15 +67,15 @@ func defaultNoTunnel() bool {
 }
 
 func defaultKeygenURL() string {
```
Comment (cihangir) [inline go/src/koding/klient/main.go]: since we have the struct we dont need helper functions

**Comment by cihangir** [comment]
could you please rebase too?

**Review by cihangir — APPROVED** [review]
